Add LegacySidecarContainers feature gate

gjkim42 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This adds the LegacySidecarContainers feature gate that enables the legacy code path that predates the SidecarContainers feature to safely remove the code.

This temporary feature gate is disabled by default, only available in v1.33, and will be removed in v1.34.

xref: https://github.com/kubernetes/kubernetes/pull/129731/files#r1923869901

/cc @SergeyKanzhelev @matthyx 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Introduced the `LegacySidecarContainers` feature gate enabling the legacy code path that predates the `SidecarContainers` feature. This temporary feature gate is disabled by default, only available in v1.33, and will be removed in v1.34.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-e2e-containerd-features
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

sftim commented:
We try not to ask people to, eg
> If you want to use legacy sidecars, disable the `DisableLegacySidecarContainers` feature gate.

Instead, I recommend having an insta-deprecated feature gate named `LegacySidecarContainers`. The _feature_ being removed is support for LegacySidecarContainers; we're not removing support for [the ability to disable legacy sidecar containers] because the subsequent Kubernetes release will continue not have legacy sidecar support enabled.

There are some feature gates matching `Disable*`, but I'm trying to avoid shipping more of them. Similarly we don't name features gates `EnableImageVolume`, `EnableNFTablesProxyMode` etc; the feature gate is `ImageVolume`, `NFTablesProxyMode` etc and you can choose whether you enable or disable it.

---

Also, we prefer to put feature gate names within backticks inside release notes.

gjkim42 commented on pkg/features/kube_features.go:
```diff
@@ -185,6 +185,14 @@ const (
+	// Disable legacy code path in pkg/kubelet/kuberuntime that predates the
+	// SidecarContainers feature. This temporary feature gate is enabled by
+	// default and intended to safely remove the redundant code path. This is
+	// only available in v1.33 and will be removed in v1.34.
```
The point of this feature is to remove the redundant code path so that we no longer have to maintain the two code paths with the same functionality as soon as possible.

If we choose to start the feature gate as `Deprecated`, is it possible to remove it in the next release?

toVersus replied:
There was a similar discussion at https://github.com/kubernetes/kubernetes/pull/129713#discussion_r1949506271.

@liggitt 
How should we handle the feature gate in this specific case?

gjkim42 commented:
> Instead, I recommend having an insta-deprecated feature gate named `LegacySidecarContainers`. The _feature_ being removed is support for legacy sidecar containers; we're not removing support for [the ability to disable legacy sidecar containers] because the subsequent Kubernetes release will continue not have legacy sidecar support enabled.

I plan to name this feature `LegacySidecarContainers`, disable it by default, mark it as deprecated, and remove it in the next release.
What do you think?
@SergeyKanzhelev

gjkim42 commented:
updated the code and PR description.

gjkim42 committed:
Add LegacySidecarContainers feature gate

This adds LegacySidecarContainers feature gate that enables the legacy
code path that predates the SidecarContainers feature to safely remove
the code.

This temporary feature gate is disabled by default, only available in
v1.33, and will be removed in v1.34.

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

thank you for the cleanup
